feat: span processor api refactor

paullegranddc opened the pull request:
# Design discussion issue (if applicable) #

https://github.com/open-telemetry/opentelemetry-rust/issues/2940

# Changes

This PR refactors the SpanProcessor API.

The API currently has the following issues:
1. It's not possible to read a span fields in a processor without calling `export_data`, which completely copies the data
2. The `SpanProcessor::on_end` API means owned span data has to be passed to each span processor, even if it does nothing with it.

One way that was explored to fix number 2. was to pass a `&mut Span` to `on_end` do processor can use std::mem::take to consume data without copying, but this is not ideal for at least two reasons:
* The otel sdk spec specifically forbids mutation of the span in `on_end`, and modifications done in one processor should not impact data passed to other processors
*  If there are multiple processors, and any processor but the last one grabs the data and leaves the span empty, this will cause bugs dependent on span processor ordering...

This PR thus proposes the following changes changes to the API:

### Introduce a `ReadableSpan` trait, implemented for SdkSpan. 

This trait define getters for span fields so we can read the span without copying or consuming data. This is inspired by other Otel libraries such as [Java](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/ReadableSpan.java) which has a similar interface.

### Modify `on_end` to pass it a `FinishedSpan`.
`FinishedSpan` is a new abstraction around a span that has been closed. It implements the `ReadableSpan` trait so we can read from it without consuming the data.
if a span processor needs to get an owned SpanData on it. This will do one of two things:
  * if this is the last span processor that will be invoked, it gives ownership of the span data without copying
  * otherwise it copies the span data
 
  This design
  1. allows not copying in span processors that don't need owned data during `on_end`
  2. doesn't expose mutable spans in `on_end`
  3. eludes the span data copy if the exporter is the last span processor (which is usually the case).
 
# Benchmarking

A benchmark running creating a span and dropping it, invoking the `on_start`, `on_ending` and `on_end` methods for a varying number of span processors has been added.

```
Baseline - main branch:

SpanProcessorApi/0_processors
    time:   [339.66 ns 340.56 ns 341.47 ns]
SpanProcessorApi/1_processors
    time:   [373.10 ns 374.36 ns 375.60 ns]
SpanProcessorApi/2_processors
    time:   [803.10 ns 804.99 ns 807.03 ns]
SpanProcessorApi/4_processors
    time:   [1.2096 µs 1.2137 µs 1.2179 µs]

Candidate - paullegranddc:paullgdc/sdk/span_processor_api_refactor:

SpanProcessorApi/0_processors
    time:   [385.15 ns 386.14 ns 387.25 ns]
SpanProcessorApi/1_processors
    time:   [385.73 ns 387.17 ns 388.85 ns]
SpanProcessorApi/2_processors
    time:   [384.84 ns 385.66 ns 386.50 ns]
SpanProcessorApi/4_processors
    time:   [386.78 ns 388.17 ns 389.58 ns]
```

On main, the cost of a span lifecycle is proportional to the number of span processors, because the `Clone` operation on span data is as expensive as creating the span.

After this PR, since span processors don't clone data unless they need it, the cost is constant regardless of the number of span processors.


## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

paullegranddc committed:
feat: add a readable span interface

Currently there is no way to read the span data without cloning it.
This causes performance issues when the span for span
processors that need to read the span data.

This commit introduces a new trait `ReadableSpan` in the sdk
implemented by SDK Spans that allows to read the span data without
cloning it.

paullegranddc committed:
feat: add an experimental on_ending api

This API allows to mutations of the span when it is ending.
It's marked as on development in the spec, but it is useful for
span obfuscation for example, which needs to done after
attributes can added to the span anymore.

paullegranddc committed:
fix: formatting and clippy

paullegranddc committed:
Merge branch 'main' into paullgdc/sdk/span_processor_api_refactor

paullegranddc committed:
fix: some SpanProcessor calls were not finished

paullegranddc committed:
fix: lint

paullegranddc committed:
fix: lint clippy

paullegranddc committed:
Merge branch 'main' into paullgdc/sdk/span_processor_api_refactor

scottgerring commented on opentelemetry-sdk/src/trace/span.rs:
```diff
@@ -198,43 +198,53 @@ impl opentelemetry::trace::Span for Span {
+            Some(data) => data,
+            None => return,
+        };
+        let span_context: SpanContext =
```
It would be helpful to comment around this bit to make it easier to re-grok -- "take ownership of the span_context leaving an empty context in its place; this saves us a clone" or so, right?

scottgerring commented on opentelemetry-sdk/src/trace/span.rs:
```diff
@@ -198,43 +198,53 @@ impl opentelemetry::trace::Span for Span {
-            None => return,
+        match self.data {
+            Some(ref mut data) => {
+                // ensure end time is set via explicit end or implicitly on drop
```
And general observation - this seems like it should be _faster_💨  - do we have some benchmarks covering we can show deltas from?

paullegranddc replied:
> Could we lift this comment up to the method itself as it is describing when the entire method is invoked?

Done

scottgerring replied:
@paullegranddc whoops, I mean like this above: 
```rust
// Ensure end time is set explicitly, or implicitly when Span is dropped
fn ensure_ended_and_exported()
```

scottgerring commented on opentelemetry-sdk/src/trace/span.rs:
```diff
@@ -246,6 +256,282 @@ impl Drop for Span {
+/// If `consume`` is never called, the on_ending method will not perform any copy of
+/// the span data.
+///
+/// Calling any `ReadableSpan` method on the `FinishedSpan` will panic if the span data
```
We should probably error out through the internal logging macros here and not panic. We try to avoid panic-ing wherever we can.

paullegranddc replied:
This does not panic anymore

scottgerring commented on opentelemetry-sdk/src/trace/span.rs:
```diff
@@ -246,6 +256,282 @@ impl Drop for Span {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut fmt = f.debug_struct("FinishedSpan");
+        if self.is_consummed {
+            fmt.field("consummed", &self.is_consummed);
```
```suggestion
            fmt.field("consumed", &self.is_consumed);
```

scottgerring commented on opentelemetry-sdk/src/trace/span.rs:
```diff
@@ -246,6 +256,282 @@ impl Drop for Span {
+pub struct FinishedSpan {
+    span: Option<crate::trace::SpanData>,
+    is_last_processor: bool,
+    is_consummed: bool,
```
```suggestion
    is_consumed: bool,
```

scottgerring reviewed (commented):
Hey @paullegranddc thanks for opening the PR!
I've had a quick zoom through now and have some comments and queries.

This feels like a sensible way of addressing the problems you mentioned (and highlighted in the examples 👍 ). Of course it will break the existing `on_end`, but as we're not stable here, yet and there's a clear upshot, from my perspective that seems reasonable. I think @cijothomas will also have helpful opinions here - Cijo? 

Adding `on_ending` seems like a no brainer. 

It would be great to look at some big downstream projects (`tracing-opentelemetry` comes to mind), and check if they are impacted. It would also be great to have some perf numbers from benchmarks touching the path through `ensure_ended_and_exported` which you may already have handy?

paullegranddc committed:
feat: add span processor benchmarks

Adds a benchmark creating a span and dropping it, without an exporter.
The goal is to estimate the cost of the SDK running without the exporter

paullegranddc commented on examples/tracing-http-propagator/src/server.rs:
```diff
@@ -104,6 +110,101 @@ async fn router(
+        else {
+            return;
+        };
+        let mut counts = self.0.lock().unwrap();
```
Yes, this panics if the mutex has been poisoned, which means a panic has happened in another thread that was holding the mutex.

Usually the best course of action is to propagate the panic because the state of the program is now uncertain (invariants that were protected by the mutex might not be true anymore)

paullegranddc committed:
fix: remove unised import

paullegranddc commented on opentelemetry-sdk/benches/batch_span_processor.rs:
```diff
@@ -4,10 +4,10 @@ use opentelemetry::trace::{
     SpanContext, SpanId, SpanKind, Status, TraceFlags, TraceId, TraceState,
```
Yes, I've run the existing `BatchSpanProcessor` benchmarks.
With `experimental_span_processor_on_ending` not enabled, the benchmark show no detectable changes in performance.

If the `experimental_span_processor_on_ending` feature is enabled, there is a slight cost (5%) which I believe comes from the fact we have to clone the tracer field, because we have to iterate over the span processor list through a shared reference, while passing the span mutably.  Thus we have to clone the tracer associated with the span so we can split the ownership (it's in an `Arc` cell so this is not that expensive)

scottgerring replied:
Nice! This is exactly what I was hoping to see :D

paullegranddc committed:
fix: typos, so many typos...

paullegranddc committed:
fix: insert count as number instead of string in example

paullegranddc committed:
fix: make on_ending contract more explicit

paullegranddc committed:
fix: on_ending comment

paullegranddc commented on opentelemetry-sdk/src/trace/span_processor.rs:
```diff
@@ -79,11 +79,24 @@ pub trait SpanProcessor: Send + Sync + std::fmt::Debug {
     /// not block or throw exceptions.
     fn on_start(&self, span: &mut Span, cx: &Context);
+    #[cfg(feature = "experimental_span_processor_on_ending")]
+    /// `on_ending` is called when a `Span` is ending. The en timestampe has already
```
I reformulated the comment, it should be a bit clearer

paullegranddc committed:
nit: move span endtime logic to callers of ensure_ended_and_exported

paullegranddc committed:
fix: remove panic in FinishedSpan

paullegranddc committed:
fix: add tests and fix clippy

paullegranddc committed:
Merge branch 'main' into paullgdc/sdk/span_processor_api_refactor

paullegranddc committed:
nit: use None rather than empty string when no recording

paullegranddc committed:
nit: add test coverage for ReadableSpan implementation

paullegranddc committed:
fix: add non panicking version of FinishedSpan::consume

paullegranddc committed:
fix: test non panicking behavior

paullegranddc committed:
fix: remove panics in finished span reads

paullegranddc committed:
fix: remove no longer relevant comment

paullegranddc committed:
nit: add comment explaining why we grab the data from the span mutably

paullegranddc committed:
fix: rename ensure_ended_and_exported to end_and_export

paullegranddc committed:
fix: example return instead of panic on posined lock

scottgerring reviewed (approved):
This LGTM modulo the current builds passing. @cijothomas wdyt?

TommyCpp commented on opentelemetry-sdk/src/trace/span.rs:
```diff
@@ -192,57 +192,366 @@ impl opentelemetry::trace::Span for Span {
+
+impl FinishedSpan {
+    /// Creates a new `FinishedSpan` with the given span data.
+    pub fn new(span_data: crate::trace::SpanData) -> Self {
```
Should we implement a `From` between those two?

TommyCpp replied:
Not suppose used by end users we can add one if needed in the future

scottgerring commented on opentelemetry-sdk/src/trace/span.rs:
```diff
@@ -192,57 +192,366 @@ impl opentelemetry::trace::Span for Span {
+            opentelemetry::otel_error!(name: "FinishedSpan.ConsumeTwice", message = "consume called twice on FinishedSpan in the same span processor");
+        }
+        self.try_consume()
+            .expect("Span data has already been consumed")
```
Should we be calling expect at all? This feels right on the line with "let's not panic in the library code", as it highlights an issue with a span processor

paullegranddc replied:
My philosophy on this is that in the end, if this panics it's because of an API misuse. Not because of wrong assumptions on the data manipulated by the program which I understand coding defensively against.
As someone implementing a span processor, I'd rather crash immediately if I misuse the API which is surely going to be caught in my tests.

Having a `try_consume` method gives you the freedom to handle the misuse if you want, but in practice as someone writing code against this API I'd rather not handling this mode of failure makes sense.

paullegranddc committed:
revert: remove on_ending API

TommyCpp committed:
Merge branch 'main' into paullgdc/sdk/span_processor_api_refactor

TommyCpp committed:
Merge branch 'main' into paullgdc/sdk/span_processor_api_refactor

paullegranddc committed:
fix: make consume non panicking

TommyCpp committed:
Merge branch 'main' into paullgdc/sdk/span_processor_api_refactor

cijothomas committed:
Merge branch 'main' into paullgdc/sdk/span_processor_api_refactor
